### PR TITLE
FEAT-#6965: Implement '.merge()' using range-partitioning implementation

### LINKED ISSUE #6965: Implement `merge()` using range-partitioning implementation

### PR DESCRIPTION
**Pull request by dchigarev**
## What do these changes do?

This PR adds a new implementation for `merge()` using range-partitioning mechanism. The new implementation does the following:
1. Computes range-partition bins for left dataframe
2. Builds range-partitioning for left dataframe using bins computed at the first step
3. Builds range-partitioning for <i>right</i> dataframe using bins computed at the first step
4. Applies the `merge` kernel row-wise to the left dataframe and broadcasts <i>row partitions</i> of the right dataframe there

The benefit of this implementation is that we never gather the whole right dataframe in one partition ([as we do it now](https://github.com/modin-project/modin/blob/271d98b260f03b345ae2384001f2df83b98322bf/modin/core/storage_formats/pandas/query_compiler.py#L523)) but rather repartition it in a way so it would be correct to broadcast only row partitions of the right df. This implementation benefits when the right dataframe is relatively big.

One of the downsides of this implementation is that repartitioning changes the order of rows, meaning that the result is always sorted by keys at the end.

At the moment, one can only use range-partitioning impl for merge by specifying a special config variable (`cfg.RangePartitioningMerge`).

<b>Perf measurements for h2o</b>
Important note: original h2o dataset has several categorical columns with high cardinality (high amount of unique values), this is a problematic case for modin ([see 2nd point](https://github.com/modin-project/modin/issues/5722#issuecomment-1450354798)) and it so it works terribly slow with such dtypes. In the measurements below, all categorical columns were casted to strings.

<details><summary>h2o joins, 500mb data</summary>

image

</details>

<details><summary>h2o joins, 5gb data</summary>

image

</details>

- [x] first commit message and PR title follow format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #6965 
- [x] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date

### CODE REVIEW

File: modin/core/storage_formats/pandas/merge.py [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+            `suffixes` argument that was passed to ``pandas.merge()``.
+
+        Returns
+        -------
+        new_columns : pandas.Index or None
+            Columns for the result of merge. ``None`` if not enought metadata to compute.
+        new_dtypes : ModinDtypes or None
+            Dtypes for the result of merge. ``None`` if not enought metadata to compute.
+        """
+        new_columns = None
```
Comment (dchigarev) [inline modin/core/storage_formats/pandas/merge.py:251]: this logic was copied without any changes from `PandasQueryCompiler.merge` and was placed in a separate method to be reused by range-partitioning impl

File: modin/config/envvars.py [hunk 2]
```diff
@@ -770,6 +770,19 @@ def _sibling(cls) -> type[EnvWithSibilings]:
 )
 
 
+class RangePartitioningMerge(EnvironmentVariable, type=bool):
```
[the remark this answers is not in the document]
Comment (anmyachev) [reply modin/config/envvars.py:773]: > What do others think in this regard?

Why not just use one config variable (like `RangePartitioningImpl`) for all the functions that support it?

[the remark this answers is not in the document]
Comment (dchigarev) [reply modin/config/envvars.py:773]: > Why not just use one config variable (like `RangePartitioningImpl`) for all the functions that support it?

Yep, it seems like the only option. We could have then expanded [this page](https://modin.readthedocs.io/en/stable/flow/modin/experimental/range_partitioning_groupby.html), giving a short description of every operation that supports range-partitioning with tips on when to use it, and then probably link this page in our [optimization notes ](https://modin.readthedocs.io/en/stable/usage_guide/optimization_notes/index.html).

[the remark this answers is not in the document]
Comment (dchigarev) [reply modin/config/envvars.py]: Added a common variable `RangePartitioning`, let's handle the transition from `RangePartitioningGroupby` as well as changes in documentation in a separate PR

File: docs/flow/modin/experimental/index.rst [hunk 3]
```diff
@@ -15,7 +15,7 @@ and provides a limited set of functionality:
 * :doc:`xgboost <xgboost>`
 * :doc:`sklearn <sklearn>`
 * :doc:`batch <batch>`
-* :doc:`Range-partitioning GroupBy implementation <range_partitioning_groupby>`
+* :doc:`Range-partitioning implementations <range_partitioning_groupby>`
```
Comment (anmyachev) [inline docs/flow/modin/experimental/index.rst:18]:
```suggestion
* :doc:`Range-partitioning implementations: <range_partitioning_groupby>`
```

File: modin/pandas/test/dataframe/test_join_sort.py [hunk 4]
```diff
@@ -214,6 +221,10 @@ def test_join_6602():
         teams.set_index("league_abbreviation").join(abbreviations.rename("league_name"))
 
 
+@pytest.mark.skipif(
+    RangePartitioning.get() and StorageFormat.get() == "Hdk",
+    reason="Doesn't work on HDK",
```
Comment (anmyachev) [inline modin/pandas/test/dataframe/test_join_sort.py:226]: ?
```suggestion
    reason="Doesn't make sense for HDK",
```

File: modin/core/dataframe/pandas/dataframe/dataframe.py [hunk 5]
```diff
@@ -3859,6 +3859,74 @@ def _compute_new_widths():
...
 
+    def _apply_func_to_range_partitioning_broadcast(
+        self, right, func, key, new_index=None, new_columns=None, new_dtypes=None
+    ):
+        """
+        Apply `func` against two dataframes using range-partitioning implementation.
+
+        The method first builds range-partitioning for both dataframes using the data from
+        `self[key]`, after that, the it applies the `func` row-wise to the `self` frame and
+        broadcasts row-parts of `right` to the `self`.
```
Comment (anmyachev) [inline modin/core/dataframe/pandas/dataframe/dataframe.py:3870]:
```suggestion
        `self[key]`, after that, it applies `func` row-wise to `self` frame and
        broadcasts row-parts of `right` to `self`.
```

File: modin/core/dataframe/pandas/partitioning/partition_manager.py [hunk 6]
```diff
@@ -1736,6 +1737,9 @@ def shuffle_partitions(
             An object implementing the functions that we will be using to perform this shuffle.
         final_shuffle_func : Callable(pandas.DataFrame) -> pandas.DataFrame
             Function that shuffles the data within each new partition.
+        right_partitions : np.ndarray, optional
+            Partitions to broadcast to the `self` partitions. If specified, the method builds range-partitioning
+            for the `right_partitions` basing on bins calculated for `partitions`, then performs broadcasting.
```
Comment (anmyachev) [inline modin/core/dataframe/pandas/partitioning/partition_manager.py:1742]:
```suggestion
            Partitions to broadcast to `self` partitions. If specified, the method builds range-partitioning
            for `right_partitions` basing on bins calculated for `partitions`, then performs broadcasting.
```

File: modin/core/dataframe/pandas/partitioning/partition_manager.py [hunk 7]
```diff
@@ -1774,18 +1778,57 @@ def shuffle_partitions(
...
-            # We need to convert every partition that came from the splits into a full-axis column partition.
-            new_partitions = [
+
+            if right_partitions is None:
+                # We need to convert every partition that came from the splits into a full-axis column partition.
+                return np.array(
+                    [
+                        [
+                            cls._column_partitions_class(
+                                row_partition, full_axis=False
```
Comment (anmyachev) [inline modin/core/dataframe/pandas/partitioning/partition_manager.py:1788]: Why `full_axis=False`? Contradicts the comment

File: modin/core/storage_formats/pandas/merge.py [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+
+        return left.__constructor__(
+            left._modin_frame._apply_func_to_range_partitioning_broadcast(
+                right._modin_frame,
+                func=func,
+                key=on,
+                new_columns=new_columns,
+                new_dtypes=new_dtypes,
+            )
+        ).reset_index(drop=True)
```
Comment (anmyachev) [inline modin/core/storage_formats/pandas/merge.py:90]: Why?

File: docs/flow/modin/experimental/index.rst (hunk 3 above)
Comment (dchigarev) [reply docs/flow/modin/experimental/index.rst:18]: don't think a colon symbol is required here, it's just a list of links:
https://modin.readthedocs.io/en/stable/flow/modin/experimental/index.html

File: modin/core/storage_formats/pandas/merge.py (hunk 8 above)
Comment (dchigarev) [reply modin/core/storage_formats/pandas/merge.py:90]: added the following in-code comment:
```
# pandas resets the index of the result unless we were merging on an index level,
# the current implementation only supports merging on column names, so dropping
# the index unconditionally
```

File: docs/flow/modin/experimental/index.rst (hunk 3 above)
Comment (anmyachev) [reply docs/flow/modin/experimental/index.rst:18]: You have plural implementations, how are you planning to specify multiple functions/files?

Comment (dchigarev) [reply docs/flow/modin/experimental/index.rst:18]: I'm planning to rename `range_partitioning_groupby.rst -> range_partitioning.rst` and use this file to store information about all the methods that have range-partitioning implementation, then I'm planing to move that page to the [optimization notes](https://modin.readthedocs.io/en/stable/usage_guide/optimization_notes/index.html) (I want to do all this in a separate PR)

Comment (anmyachev) [reply docs/flow/modin/experimental/index.rst:18]: ok

**Review by anmyachev — APPROVED** [review]
LGTM!

File: modin/pandas/test/dataframe/test_join_sort.py [hunk 9]
```diff
@@ -236,6 +247,10 @@ def test_join_6602():
     ],
 )
 def test_merge(test_data, test_data2):
+    # RangePartitioning merge always produces sorted result, so we have to sort
```
Comment (YarShev) [inline modin/pandas/test/dataframe/test_join_sort.py:250]: sorted -> unsorted?

**Pull request merged**